README.md
sturfee Cookbook
================
TODO: Enter the cookbook description here.

e.g.
This cookbook makes your favorite breakfast sandwich.

Requirements
------------
TODO: List your cookbook requirements. Be sure to include any requirements this cookbook has on platforms, libraries, other cookbooks, packages, operating systems, etc.

e.g.
#### packages
- `toaster` - sturfee needs toaster to brown your bagel.

Attributes
----------
TODO: List your cookbook attributes here.

e.g.
#### sturfee::default
<table>
  <tr>
    <th>Key</th>
    <th>Type</th>
    <th>Description</th>
    <th>Default</th>
  </tr>
  <tr>
    <td><tt>['sturfee']['bacon']</tt></td>
    <td>Boolean</td>
    <td>whether to include bacon</td>
    <td><tt>true</tt></td>
  </tr>
</table>

Usage
-----
#### sturfee::default
TODO: Write usage instructions for each cookbook.

e.g.
Just include `sturfee` in your node's `run_list`:

```json
{
  "name":"my_node",
  "run_list": [
    "recipe[sturfee]"
  ]
}
```

Contributing
------------
TODO: (optional) If this is a public cookbook, detail the process for contributing. If this is a private cookbook, remove this section.

e.g.
1. Fork the repository on Github
2. Create a named feature branch (like `add_component_x`)
3. Write your change
4. Write tests for your change (if applicable)
5. Run the tests, ensuring they all pass
6. Submit a Pull Request using Github

License and Authors
-------------------
Authors: TODO: List authors


CHANGELOG.md
sturfee CHANGELOG
=================

0.2.0
-----
Will be TDD and restart services at the end of the run.

0.1.0
-----
It works, but is not TDD and doesn't restart services at the end of chef run.


spec/spec_helper.rb
require 'chefspec'
# require 'chefspec/berkshelf'
require 'chefspec/librarian'

ChefSpec::Coverage.start!

def puts! args, label=""
  puts "+++ +++ #{label}"
  puts args.inspect
end


spec/recipes/apiserv_spec.rb
require 'spec_helper'

describe 'sturfee::apiserv' do
  let(:chef_run) do
    ChefSpec::SoloRunner.new do |node|
      node.set['cookbook']['attribute'] = 'hello'
      node.set['roles'] = [ 'some', 'roles' ]
    end.converge("role[sturfee_apiserv]") # described_recipe
  end
  
  before :each do
    stub_search("apps", "*:*").and_return([{ :id => 'sturfee_apiserv' }])
    @service_name = "sturfee_apiserv-app"
    @projects_dir = "/home/ubuntu/projects"
    stub_data_bag_item("apps", "sturfee_apiserv").and_return(
      'id' => 'sturfee_apiserv',
      'force' => { '_default' => true },
      'type' => { 'sturfee_apiserv' => [ 'apiserv' ] },
      'packages' => {'wget' => ''},
      'user' => { "_default" => "ubuntu" },
      'revision' => { "_default" => "master" }
    )
  end
  
  it 'installs app packages' do
    expect(chef_run).to install_package("wget")
  end
    
  it 'unlocks apiserv file' do
  end

  it 'creates letsencrypt/<DOMAIN> dir' do
  end

  it 'renders deploy wrapper' do
    expect(chef_run).to render_file( "#{@projects_dir}/deploy-ssh-wrapper" )
  end

  it 'gets the executable server' do
  end

  it 'writes static config' do
  end

  it 'writes ssl fullchain' do
  end

  it 'writes ssl privkey' do
  end

  it 'creates upstart service' do
  end
  
  it 'restarts the service at the end of the run' do
    expect(chef_run).to restart_service(@service_name)  
    false.should eql true
  end
  
end


recipes/apiserv.rb

#
# Recipe sturfee::apiserv
#
# _vp_ 20151228, 20160329
#

search(:apps).each do |any_app|
  node.roles.each do |role|
    if any_app['id'] == role
      app = data_bag_item 'apps', any_app['id']
      if app['type'][app['id']].include? 'apiserv'
        puts! "sturfee::apiserv, deploying #{role}"

        ## config
        user                = app['user'][node.chef_environment]
        projects_dir        = "/home/#{user}/projects"
        app['deploy_to']    = "#{projects_dir}/#{app['id']}"
        upstart_script_name = "#{app['id']}-app"
        domain              = app['domains'][node.chef_environment][0]
        region              = app['s3'][node.chef_environment]['region']
        s3_key              = app['s3'][node.chef_environment]['key']
        s3_secret           = app['s3'][node.chef_environment]['secret']
        s3_deploy_bucket    = app['s3'][node.chef_environment]['buckets']['deploy']


        app['packages'].each do |pkg, version|
          package pkg
        end

        execute "unlock apiserv file" do
          command "service #{upstart_script_name} stop ; sleep 3 ; pkill apiserv ; echo ok"
        end

        ##
        ## create all the directories
        ##
        directory "#{app['deploy_to']}/config/letsencrypt/live/#{domain}" do
          action :create
          recursive true
          owner user
          group user
        end

        # create deploy wrapper
        ruby_block "write_key" do
          block do
            f = ::File.open("#{projects_dir}/id_deploy", "w")
            f.print(app["deploy_key"])
            f.close
          end
          not_if do ::File.exists?("#{projects_dir}/id_deploy"); end
        end
        file "#{projects_dir}/id_deploy" do
          owner user
          group user
          mode '0600'
        end
        template "#{projects_dir}/deploy-ssh-wrapper" do
          source "deploy-ssh-wrapper.erb"
          owner user
          group user
          mode "0755"
          variables({
                      :deploy_to => projects_dir # this is only for reference of id_deploy file.
                    })
        end

        ## get the executable server
        execute 'get the executable server' do
          command %{AWS_ACCESS_KEY_ID=#{s3_key} AWS_SECRET_ACCESS_KEY=#{s3_secret} aws s3 cp s3://#{s3_deploy_bucket}/apiserv/latest/apiserv apiserv --region #{region}}
          cwd "#{app['deploy_to']}"
          notifies :restart, "service[#{upstart_script_name}]", :delayed
        end
        file "#{app['deploy_to']}/apiserv" do
          mode '0755'
        end
        # execute 'get emailTmpl.html' do
        #   command %{AWS_ACCESS_KEY_ID=#{s3_key} AWS_SECRET_ACCESS_KEY=#{s3_secret} aws s3 cp s3://#{s3_deploy_bucket}/apiserv/latest/emailTmpl.html emailTmpl.html --region #{region}}
        #   cwd "#{app['deploy_to']}"
        # end

        ## write staticserv config
        template "#{app['deploy_to']}/config/apiserv.json" do
          source "app/config/apiserv.json.erb"
          variables({
            :domain        => domain,
            :db_user       => app['mysql_db_user'][node.chef_environment],
            :db_university => app['mysql_db_university'][node.chef_environment],
            :s3            => app['s3'][node.chef_environment],
            :mailer        => app['mailer'][node.chef_environment],
            :fb            => app['facebook'][node.chef_environment]
          })
        end

        ##
        ## write ssl pem files
        ##
        ruby_block "write_ssl_fullchain" do
          block do
            f = ::File.open("#{app['deploy_to']}/config/letsencrypt/live/#{domain}/fullchain.pem", "w")
            f.print(app["letsencrypt"][node.chef_environment]['fullchain'])
            f.close
          end
        end
        file "#{app['deploy_to']}/config/letsencrypt/live/#{domain}/fullchain.pem" do
          owner user
          group user
          mode '0600'
        end
        ruby_block "write_ssl_privkey" do
          block do
            f = ::File.open("#{app['deploy_to']}/config/letsencrypt/live/#{domain}/privkey.pem", "w")
            f.print(app["letsencrypt"][node.chef_environment]['privkey'])
            f.close
          end
        end
        file "#{app['deploy_to']}/config/letsencrypt/live/#{domain}/privkey.pem" do
          owner user
          group user
          mode '0600'
        end

        ##
        ## service
        ##
        template "/etc/init/#{upstart_script_name}.conf" do
          source "etc/init/upstart_apiserv.conf.erb"
          owner  "root"
          group  "root"
          mode   "0664"
          variables(
            :app_name        => app['id'],
            :app_root        => "#{app['deploy_to']}",
            :executable_name => 'apiserv'
          )
        end
        service upstart_script_name do
          provider Chef::Provider::Service::Upstart
          supports :status => true, :restart => true
          action [ :enable, :start, :restart ]
        end

        ## restart service here
        ## I'm doing this via notify. _vp_ 20160405
        # service upstart_script_name do
        #   action [ :restart ]
        # end

      end
    end
  end
end








recipes/staticserv.rb

#
# Recipe sturfee::staticserv
#
# _vp_ 20151228, 20160329
#

search(:apps).each do |any_app|
  node.roles.each do |role|
    if any_app['id'] == role
      app = data_bag_item 'apps', any_app['id']
      if app['type'][app['id']].include? 'staticserv'
        puts! "sturfee::staticserv, deploying #{role}"

        ## config
        user                = app['user'][node.chef_environment]
        projects_dir        = "/home/#{user}/projects"
        app['deploy_to']    = "#{projects_dir}/#{app['id']}"
        upstart_script_name = "#{app['id']}-app"
        domain              = app['domains'][node.chef_environment][0]
        # region              = app['s3'][node.chef_environment]['region']
        # s3_key              = app['s3'][node.chef_environment]['key']
        # s3_secret           = app['s3'][node.chef_environment]['secret']
        # s3_deploy_bucket    = app['s3'][node.chef_environment]['buckets']['deploy']
        
        app['packages'].each do |pkg, version|
          package pkg
        end

        execute "unlock staticserv file" do
          command "service #{upstart_script_name} stop ; sleep 3 ; pkill staticserv ; echo ok"
        end

        directory projects_dir do
          action :create
          recursive true
          owner user
          group user
        end

        # deploy resource
        ruby_block "write_key" do
          block do
            f = ::File.open("#{projects_dir}/id_deploy", "w")
            f.print(app["deploy_key"])
            f.close
          end
          not_if do ::File.exists?("#{projects_dir}/id_deploy"); end
        end
        file "#{projects_dir}/id_deploy" do
          owner user
          group user
          mode '0600'
        end
        template "#{projects_dir}/deploy-ssh-wrapper" do
          source "deploy-ssh-wrapper.erb"
          owner user
          group user
          mode "0755"
          variables({
                      :deploy_to => projects_dir # this is only for reference of id_deploy file.
                    })
        end
        
        #
        # deploy resource
        #
        deploy_revision app['id'] do
          revision     app['revision'][node.chef_environment]
          repository   app['repository']
          user         user
          group        user
          deploy_to    app['deploy_to']
          environment  'RAILS_ENV' => app['rack_environment']
          action        app['force'][node.chef_environment] ? :force_deploy : :deploy
          ssh_wrapper   "#{projects_dir}/deploy-ssh-wrapper" if app['deploy_key']
          shallow_clone true
          migrate       false
        end
        
        ## configure the api endpoint
        # template "#{app['deploy_to']}/current/public/js/config.js" do
        #   source "app/public/js/config.js.erb"
        #   owner user
        #   group user
        #   mode "0664"
        #   variables(
        #     :endpoint => app['api_endpoint']
        #   )
        # end
                
        service "#{app['id']}-app" do
          action :reload
        end

        ## get the executable server
        execute 'get the executable server' do
          command %{AWS_ACCESS_KEY_ID=#{app['s3_key']} AWS_SECRET_ACCESS_KEY=#{app['s3_secret']} aws s3 cp s3://#{app['s3_bucket']}/staticserv/latest/staticserv staticserv --region us-west-2}
          cwd "#{app['deploy_to']}/current"
        end
        file "#{app['deploy_to']}/current/staticserv" do
          mode '0755'
        end

        ## write staticserv config
        template "#{app['deploy_to']}/current/config/staticserv.json" do
          source "app/config/staticserv.json.erb"
          variables({
            :domain      => app['domains'][node.chef_environment][0],
            :deploy_path => "#{app['deploy_to']}/current"
          })
        end

        ##
        ## write ssl pem files
        ##
        directory "#{app['deploy_to']}/current/config/letsencrypt/live/#{domain}" do
          action :create
          recursive true
          owner user
          group user
        end
        ruby_block "write_ssl_fullchain" do
          block do
            f = ::File.open("#{app['deploy_to']}/current/config/letsencrypt/live/#{domain}/fullchain.pem", "w")
            f.print(app["letsencrypt"][node.chef_environment]['fullchain'])
            f.close
          end
        end
        file "#{app['deploy_to']}/current/config/letsencrypt/live/#{domain}/fullchain.pem" do
          owner user
          group user
          mode '0600'
        end
        ruby_block "write_ssl_privkey" do
          block do
            f = ::File.open("#{app['deploy_to']}/current/config/letsencrypt/live/#{domain}/privkey.pem", "w")
            f.print(app["letsencrypt"][node.chef_environment]['privkey'])
            f.close
          end
        end
        file "#{app['deploy_to']}/current/config/letsencrypt/live/#{domain}/privkey.pem" do
          owner user
          group user
          mode '0600'
        end

        
        ## write js config
        template "#{app['deploy_to']}/current/public/app/config.js" do
          source "app/public/app/config.js.erb"
          variables({
            :api_endpoint     => app['api_endpoint'][node.chef_environment],
            :storage_endpoint => app['storage_endpoint'][node.chef_environment],
            :publication_id   => app['publication_id'][node.chef_environment]
          })
        end

        ##
        ## service
        ##
        template "/etc/init/#{upstart_script_name}.conf" do
          source "etc/init/upstart.conf.erb"
          owner  "root"
          group  "root"
          mode   "0664"
          variables(
            :app_name       => app['id'],
            :app_root       => "#{app['deploy_to']}/current"
          )
        end
        service upstart_script_name do
          provider Chef::Provider::Service::Upstart
          supports :status => true, :restart => true
          action [ :enable, :start ]
        end

      end
    end
  end
end






